### PR TITLE
'Document list' component to replace downloads and articles, and use for search results

### LINKED ISSUE #1664: Download resources checkboxes cause console error
image
Clicking the download resources filter checkboxes cause a console error

### LINKED ISSUE #1797: Search result item
## What
- Type: **Component**

A new component is required for the search result items which have been evaluated as part of the on-site search project.
No existing component exists for the search items.

## Why

> Explain why you think this feature or content should be added to the ONS Design System.
>
> - What evidence do you have that your ONS service needs it?

- The user need is to facilitate improved findability of the ONS site content.
- The variants that exist are a refinement of existing historical patterns in use across the ONS website which have yet to be documented in the design system.
- Multiple rounds of user testing (Figma prototypes) of the new search design have been conducted and the designs have been approved by users each time.

## Supporting material

> Include links to any examples, research, prototypes or code that support your proposal.

Documented in **[Figma here](https://www.figma.com/file/kAyqXh8f02S0UXauLubSkF/Beta-release-reference?node-id=0%3A1)**.
All content type results

## Contacts

@armstrongb - interaction design
Leon Rose - user research (DST)
Gracie Lewis / Ryan Flynn - product owner (DST)

### PR DESCRIPTION
**Pull request by jrbarnes9**
### What is the context of this PR?
Created a new single 'document list' component to be used to display similar types of document links/excerpts. Intended for: 
- downloadable documents (e.g. print resources)
- news articles
- search results

Each variant of this component contains:
- Thumbnail image or placeholder (optional)
- Document item title/link
- Metadata (document type, date, file info)
- Description (optional)
- 'Featured' option, to draw attention to a single document displayed first on the page – aligned style to current 'featured' article

### What are the breaking changes?
Developers using the following patterns and components will need to replace them with the new `onsDocumentList()` macro.

#### Removed components
The new 'document list' component replaces the now deprecated components:
- articles
- downloads

#### Affected patterns
The following patterns are updated to now use the new 'document' component in replace of the deprecated components listed above:
- download resources (help users to...)
- news (pages)

Closes #1797 and fixes #1664 

### How to review
Check macro, documentation, variants and examples of the components and patterns listed above.

### CODE REVIEW

**Review by kruncher — APPROVED** [review]
Looks good to me. The only thing that I wonder about is the name of the component; i.e. `Document` vs something like `DocumentEntry` or `DocumentItem` since the component presents information about a document rather than being the actual document.

**Review by kruncher — APPROVED** [review]

File: src/components/document-list/_macro.njk [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+                    <div class="ons-document-list__item-content">
+
+                        <div class="ons-document-list__item-header{{ ' ons-document-list__item-header--reverse' if document.showMetadataFirst is defined and document.showMetadataFirst == true }}">
+
+                            {% set titleTag = params.titleTag | default("h2") %}
+
+                            <{{ titleTag }} class="ons-document-list__item-title ons-u-fs-m ons-u-mt-no ons-u-mb-xs">
+                                <a href="{{ document.url }}">{{ document.title }}
+                                {%- if document.metadata.file is defined and document.metadata.file %}
```
Comment (rmccar) [inline src/components/document-list/_macro.njk:33]: Need an if around this section including the fileType stuff that checks if `document.metadata` exists

File: src/components/document-list/_macro.njk [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                                                {% set prefixClasses = "ons-u-fw-b" if document.metadata.date.showPrefix is defined and document.metadata.date.showPrefix == true else "ons-u-vh" %}
+                                                <span class="{{ prefixClasses }}">{{ document.metadata.date.prefix | default("Published") }}: </span>
+                                            {%- endif -%}
+                                            {% if document.metadata.date.iso is defined and document.metadata.date.iso -%}
+                                                <time datetime="{{ document.metadata.date.iso }}">{{ document.metadata.date.short }}</time>
+                                            {%- endif %}
+                                        </li>
+                                    {% endif %}
+
+                                    {%- if document.metadata.type.text is defined and document.metadata.type.text %}
```
Comment (rmccar) [inline src/components/document-list/_macro.njk:65]:
```suggestion
                                    {%- if document.metadata.type is defined and document.metadata.type and document.metadata.type.text is defined and document.metadata.type.text %}
```

File: src/components/document-list/_macro.njk [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                                            {%- if document.metadata.type.url is defined and document.metadata.type.url %}
+                                                </a>
+                                            {% endif %}
+                                            {%- if document.metadata.type.ref is defined and document.metadata.type.ref %}
+                                                <span>{{ document.metadata.type.ref }}</span>
+                                            {% endif %}
+                                        </li>
+                                    {% endif %}
+
+                                    {%- if document.metadata.file.fileType is defined and document.metadata.file.fileType %}
```
Comment (rmccar) [inline src/components/document-list/_macro.njk:80]:
```suggestion
                                    {%- if document.metadata.file is defined and document.metadata.file and document.metadata.file.fileType is defined and document.metadata.file.fileType %}
```

**Review by rmccar — CHANGES_REQUESTED** [review]
Just these few things, other than that looks good

**Review by kruncher — DISMISSED** [review]

**Review by rmccar — APPROVED** [review]

**Comment by boxadesign** [comment]
image

Pagination layout issue on News pattern

**Comment by jrbarnes9** [comment]
> image
> 
> Pagination layout issue on News pattern


Yeah I raised a bug for this #2024

**Review by boxadesign — APPROVED** [review]

**Pull request merged**